skiko/src/jvmMain/cpp/windows/directXRedrawer.cc

```diff
@@ -0,0 +1,197 @@
+        {
+            return false;
+        }
+        gr_cp<IDXGIAdapter1> hardwareAdapter;
+        defineHardwareAdapter(factory.get(), &hardwareAdapter);
+        gr_cp<ID3D12Device> device;
+        if (!SUCCEEDED(D3D12CreateDevice(hardwareAdapter.get(), D3D_FEATURE_LEVEL_11_0, IID_PPV_ARGS(&device))))
+        {
+            return false;
+        }
+
+        // Create the command queue
+        gr_cp<ID3D12CommandQueue> queue;
+        D3D12_COMMAND_QUEUE_DESC queueDesc = {};
+        queueDesc.Flags = D3D12_COMMAND_QUEUE_FLAG_NONE;
+        queueDesc.Type = D3D12_COMMAND_LIST_TYPE_DIRECT;
+
+        if (!SUCCEEDED(device->CreateCommandQueue(&queueDesc, IID_PPV_ARGS(&queue))))
+        {
+            return false;
+        }
+
+        d3dDevice->backendContext.fAdapter = hardwareAdapter;
+        d3dDevice->backendContext.fDevice = device;
+        d3dDevice->backendContext.fQueue = queue;
+
+        d3dDevice->device = device;
+        d3dDevice->queue = queue;
+
+        return reinterpret_cast<jlong>(d3dDevice);
+    }
+
+    JNIEXPORT void JNICALL Java_org_jetbrains_skiko_redrawer_Direct3DRedrawer_createSwapChain(
+        JNIEnv *env, jobject redrawer, jlong windowHandle, jlong devicePtr)
+    {
+        DirectXDevice *d3dDevice = reinterpret_cast<DirectXDevice *>(static_cast<uintptr_t>(devicePtr));
+
+        // Make the swapchain
+        HWND fWindow = (HWND)windowHandle;
+
+        gr_cp<IDXGIFactory4> factory;
+        GR_D3D_CALL_ERRCHECK(CreateDXGIFactory2(0, IID_PPV_ARGS(&factory)));
+
+        DXGI_SWAP_CHAIN_DESC1 swapChainDesc = {};
+        swapChainDesc.BufferCount = 2;
+        swapChainDesc.Format = DXGI_FORMAT_R8G8B8A8_UNORM;
+        swapChainDesc.BufferUsage = DXGI_USAGE_RENDER_TARGET_OUTPUT;
+        swapChainDesc.SwapEffect = DXGI_SWAP_EFFECT_FLIP_DISCARD;
+        swapChainDesc.SampleDesc.Count = 1;
+
+        DXGI_SWAP_CHAIN_FULLSCREEN_DESC swapChainFSDesc = {};
+        swapChainFSDesc.Windowed = TRUE;
+
+        gr_cp<IDXGISwapChain1> swapChain;
+        GR_D3D_CALL_ERRCHECK(factory->CreateSwapChainForHwnd(d3dDevice->queue.get(), fWindow, &swapChainDesc, &swapChainFSDesc, nullptr, &swapChain));
+        GR_D3D_CALL_ERRCHECK(swapChain->QueryInterface(IID_PPV_ARGS(&d3dDevice->swapChain)));
+        GR_D3D_CALL_ERRCHECK(d3dDevice->device->CreateFence(0, D3D12_FENCE_FLAG_NONE, IID_PPV_ARGS(&d3dDevice->fence)));
+        d3dDevice->fenceEvent = CreateEvent(nullptr, FALSE, FALSE, nullptr);
+    }
+
+    JNIEXPORT void JNICALL Java_org_jetbrains_skiko_redrawer_Direct3DRedrawer_resizeBuffers(
+        JNIEnv *env, jobject redrawer, jlong devicePtr, jint width, jint height)
+    {
+        DirectXDevice *d3dDevice = reinterpret_cast<DirectXDevice *>(static_cast<uintptr_t>(devicePtr));
+        if (!d3dDevice->isSizeEqualTo(width, height))
+        {
+            GR_D3D_CALL_ERRCHECK(d3dDevice->swapChain->ResizeBuffers(2, width, height, DXGI_FORMAT_R8G8B8A8_UNORM, 0));
```

Comment:
What is `2` here?

Reply:
The number of buffers. One is the front buffer, the second is the back one. They change during a swap operation.

Reply:
Created const value `BuffersCount`.